zetaclient/evm_client.go

```diff
@@ -626,47 +614,99 @@ func (ob *EVMChainClient) observeOutTx() {
 	}
 }
 
-// return the status of txHash
-// receipt nil, err non-nil: txHash not found
-// receipt nil, err nil: txHash receipt recorded, but may not be confirmed
-// receipt non-nil, err nil: txHash confirmed
-func (ob *EVMChainClient) queryTxByHash(txHash string, nonce uint64) (*ethtypes.Receipt, *ethtypes.Transaction, error) {
-	logger := ob.logger.ObserveOutTx.With().Str("txHash", txHash).Uint64("nonce", nonce).Logger()
-	if ob.outTXConfirmedReceipts[ob.GetTxID(nonce)] != nil && ob.outTXConfirmedTransaction[ob.GetTxID(nonce)] != nil {
-		return nil, nil, fmt.Errorf("queryTxByHash: txHash %s receipts already recorded", txHash)
-	}
+// SetPendingTx sets the pending transaction in memory
+func (ob *EVMChainClient) SetPendingTx(nonce uint64, transaction *ethtypes.Transaction) {
+	ob.Mu.Lock()
+	ob.outTxPendingTransaction[ob.GetTxID(nonce)] = transaction
+	ob.Mu.Unlock()
+}
+
+// GetPendingTx gets the pending transaction from memory
+func (ob *EVMChainClient) GetPendingTx(nonce uint64) *ethtypes.Transaction {
+	ob.Mu.Lock()
+	transaction := ob.outTxPendingTransaction[ob.GetTxID(nonce)]
+	ob.Mu.Unlock()
+	return transaction
+}
+
+// SetTxNReceipt sets the receipt and transaction in memory
+func (ob *EVMChainClient) SetTxNReceipt(nonce uint64, receipt *ethtypes.Receipt, transaction *ethtypes.Transaction) {
+	ob.Mu.Lock()
+	delete(ob.outTxPendingTransaction, ob.GetTxID(nonce)) // remove pending transaction, if any
+	ob.outTXConfirmedReceipts[ob.GetTxID(nonce)] = receipt
+	ob.outTXConfirmedTransaction[ob.GetTxID(nonce)] = transaction
+	ob.Mu.Unlock()
+}
+
+// getTxNReceipt gets the receipt and transaction from memory
+func (ob *EVMChainClient) GetTxNReceipt(nonce uint64) (*ethtypes.Receipt, *ethtypes.Transaction) {
+	ob.Mu.Lock()
+	receipt := ob.outTXConfirmedReceipts[ob.GetTxID(nonce)]
+	transaction := ob.outTXConfirmedTransaction[ob.GetTxID(nonce)]
+	ob.Mu.Unlock()
+	return receipt, transaction
+}
+
+// hasTxConfirmed returns true if there is a confirmed tx for 'nonce'
+func (ob *EVMChainClient) hasTxConfirmed(nonce uint64) bool {
+	ob.Mu.Lock()
+	confirmed := ob.outTXConfirmedReceipts[ob.GetTxID(nonce)] != nil && ob.outTXConfirmedTransaction[ob.GetTxID(nonce)] != nil
+	ob.Mu.Unlock()
+	return confirmed
+}
+
+// confirmTxByHash checks if a txHash is confirmed and saves transaction and receipt in memory
+// returns nil if confirmed or error otherwise
+func (ob *EVMChainClient) confirmTxByHash(txHash string, nonce uint64) error {
 	ctxt, cancel := context.WithTimeout(context.Background(), 3*time.Second)
 	defer cancel()
 
+	// query transaction
+	transaction, isPending, err := ob.evmClient.TransactionByHash(ctxt, ethcommon.HexToHash(txHash))
+	if err != nil {
+		return err
+	}
+	if transaction == nil { // should not happen
+		log.Error().Msgf("confirmTxByHash: transaction is nil for txHash %s nonce %d", txHash, nonce)
+		return fmt.Errorf("transaction is nil")
```

Comment:
We should return txHash and nonce in error

Reply:
[commit](https://github.com/zeta-chain/node/pull/1518/commits/0b42c7b456b475c9cf0e40aa62a2875d5da966ff)